storage/badger/cache_b.go

```diff
@@ -0,0 +1,152 @@
+package badger
+
+import (
+	"errors"
+	"fmt"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+func withLimitB[K comparable, V any](limit uint) func(*CacheB[K, V]) {
+	return func(c *CacheB[K, V]) {
+		c.limit = limit
+	}
+}
+
+type storeFuncB[K comparable, V any] func(key K, val V) func(storage.BadgerReaderBatchWriter) error
+
+func withStoreB[K comparable, V any](store storeFuncB[K, V]) func(*CacheB[K, V]) {
+	return func(c *CacheB[K, V]) {
+		c.store = store
+	}
+}
+
+func noStoreB[K comparable, V any](_ K, _ V) func(storage.BadgerReaderBatchWriter) error {
+	return func(tx storage.BadgerReaderBatchWriter) error {
+		return fmt.Errorf("no store function for cache put available")
+	}
+}
+
+// nolint: unused
+func noopStoreB[K comparable, V any](_ K, _ V) func(storage.BadgerReaderBatchWriter) error {
+	return func(tx storage.BadgerReaderBatchWriter) error {
+		return nil
+	}
+}
+
+type retrieveFuncB[K comparable, V any] func(key K) func(storage.Reader) (V, error)
+
+func withRetrieveB[K comparable, V any](retrieve retrieveFuncB[K, V]) func(*CacheB[K, V]) {
+	return func(c *CacheB[K, V]) {
+		c.retrieve = retrieve
+	}
+}
+
+func noRetrieveB[K comparable, V any](_ K) func(storage.Reader) (V, error) {
+	return func(tx storage.Reader) (V, error) {
+		var nullV V
+		return nullV, fmt.Errorf("no retrieve function for cache get available")
+	}
+}
+
+type CacheB[K comparable, V any] struct {
+	metrics  module.CacheMetrics
+	limit    uint
+	store    storeFuncB[K, V]
+	retrieve retrieveFuncB[K, V]
+	resource string
+	cache    *lru.Cache[K, V]
+}
+
+func newCacheB[K comparable, V any](collector module.CacheMetrics, resourceName string, options ...func(*CacheB[K, V])) *CacheB[K, V] {
+	c := CacheB[K, V]{
+		metrics:  collector,
+		limit:    1000,
+		store:    noStoreB[K, V],
+		retrieve: noRetrieveB[K, V],
+		resource: resourceName,
+	}
+	for _, option := range options {
+		option(&c)
+	}
+	c.cache, _ = lru.New[K, V](int(c.limit))
+	c.metrics.CacheEntries(c.resource, uint(c.cache.Len()))
+	return &c
+}
+
+// IsCached returns true if the key exists in the cache.
+// It DOES NOT check whether the key exists in the underlying data store.
+func (c *CacheB[K, V]) IsCached(key K) bool {
+	return c.cache.Contains(key)
+}
+
+// Get will try to retrieve the resource from cache first, and then from the
+// injected. During normal operations, the following error returns are expected:
+//   - `storage.ErrNotFound` if key is unknown.
+func (c *CacheB[K, V]) Get(key K) func(storage.Reader) (V, error) {
+	return func(tx storage.Reader) (V, error) {
```

Comment:
```suggestion
	return func(reader storage.Reader) (V, error) {
```
Far from the highest priority, but it would be great if we could replace "transaction" terminology with "reader" or "writer" terminology where we notice it in new code.